File: EticaretJava/app/src/main/java/com/example/eticaretjava/viewmodel/AuthViewModel.java
package com.example.eticaretjava.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.eticaretjava.model.User.UserDto;
import com.example.eticaretjava.repo.RepositoriesImpl;
import com.example.eticaretjava.repo.ResultCallback;

public class AuthViewModel extends ViewModel {

    public static class AuthState {
        public boolean inFlight = false;
        public String error = null;
        public boolean loggedIn = false;
        public boolean registered = false;
        public UserDto me = null;
    }

    private final MutableLiveData<AuthState> state = new MutableLiveData<>(new AuthState());
    private final RepositoriesImpl.AuthRepositoryImpl repo;

    public AuthViewModel(RepositoriesImpl.AuthRepositoryImpl repo) {
        this.repo = repo;
    }

    public LiveData<AuthState> getState() {
        return state;
    }

    private AuthState copyState() {
        AuthState cur = state.getValue() != null ? state.getValue() : new AuthState();
        AuthState n = new AuthState();
        n.inFlight = cur.inFlight;
        n.error = cur.error;
        n.loggedIn = cur.loggedIn;
        n.registered = cur.registered;
        n.me = cur.me;
        return n;
    }

    public void login(String email, String password) {
        AuthState s = copyState();
        s.inFlight = true;
        s.error = null;
        s.loggedIn = false;
        state.setValue(s);

        repo.login(email, password, new ResultCallback<Void>() {
            @Override
            public void onSuccess(Void data) {
                AuthState ns = copyState();
                ns.inFlight = false;
                ns.loggedIn = true;
                state.postValue(ns);
            }

            @Override
            public void onError(String message) {
                AuthState ns = copyState();
                ns.inFlight = false;
                ns.error = message != null ? message : "Login failed";
                state.postValue(ns);
            }
        });
    }

    public void register(String name, String email, String password) {
        AuthState s = copyState();
        s.inFlight = true;
        s.error = null;
        s.registered = false;
        state.setValue(s);

        repo.register(name, email, password, new ResultCallback<Void>() {
            @Override
            public void onSuccess(Void data) {
                AuthState ns = copyState();
                ns.inFlight = false;
                ns.registered = true;
                state.postValue(ns);
            }

            @Override
            public void onError(String message) {
                AuthState ns = copyState();
                ns.inFlight = false;
                ns.error = message != null ? message : "Register failed";
                state.postValue(ns);
            }
        });
    }

    public void clearError() {
        AuthState ns = copyState();
        ns.error = null;
        state.setValue(ns);
    }

    public void clearRegistered() {
        AuthState ns = copyState();
        ns.registered = false;
        state.setValue(ns);
    }
}

File: Eticaretjetpack/app/src/main/java/com/example/eticaretjetpack/viewmodel/AuthViewModel.kt
package com.example.eticaretjetpack.viewmodel

// ui/auth/AuthViewModel.kt
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.eticaretjetpack.model.UserDto
import com.example.eticaretjetpack.repo.AuthRepositoryImpl
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
class AuthViewModel(
    private val repo: AuthRepositoryImpl
) : ViewModel() {

    data class AuthState(
        val inFlight: Boolean = false,
        val error: String? = null,
        val loggedIn: Boolean = false,
        val registered: Boolean = false,   // ✅ eklendi
        val me: UserDto? = null
    )

    private val _state = MutableStateFlow(AuthState())
    val state: StateFlow<AuthState> = _state

    fun login(email: String, password: String) = viewModelScope.launch {
        _state.update { it.copy(inFlight = true, error = null, loggedIn = false) }

        repo.login(email, password)
            .onSuccess {
                _state.update { it.copy(inFlight = false, loggedIn = true) }
            }
            .onFailure { e ->
                _state.update { it.copy(inFlight = false, error = e.message ?: "Login failed") }
            }
    }

    fun register(name: String, email: String, password: String) = viewModelScope.launch {
        _state.update { it.copy(inFlight = true, error = null, registered = false) }

        repo.register(name, email, password)
            .onSuccess {
                // ✅ register başarılı → loggedIn yapmıyoruz, sadece registered=true
                _state.update { it.copy(inFlight = false, registered = true) }
            }
            .onFailure { e ->
                _state.update { it.copy(inFlight = false, error = e.message ?: "Register failed") }
            }
    }

    fun clearError() = _state.update { it.copy(error = null) }

    fun clearRegistered() = _state.update { it.copy(registered = false) } // ✅ eklendi
}
